tensorflow/python/kernel_tests/conv_ops_3d_test.py

```diff
@@ -317,6 +316,33 @@ def testConv3D2x2x2FilterStride2Same(self):
         padding="SAME",
         expected=expected_output)
 
+  def _TestConv3DEmptyTensorOutputShape(self):
+    """Verifies the output shape of the Conv3D op when output tensor is empty.
+
+    Args: none
+    """
+    input_shape = [0, 112, 112, 112, 32]
```

Comment:
Shouldn't this shape be invalid? If it's all 0s then it's okay. But 0s in just some dimensions is just not possible (dim sizes should be 1s instead of 0s).

Reply:
During our internal testing, one model encountered this shape, Eigen did not error out and execution proceeded. [Intel MKL] version had cascade errors related to these corner cases. 
Do you want me to investigate why a model would create such a shape before this PR gets merged?

Reply:
That would be great. Thank you very much! 
I'm looping in @ezhulenev for comments on Eigen and grappler.

Reply:
Ok, I will do the investigation and report back.

Reply:
I've seen in multiple kernels check for zero in dimensions, though I'm also not sure that this is legal, and how that can happen.

Reply:
apassos@ usually has answers for all questions.

Reply:
Ok, could be user application bug/corner case that slipped the checks. I will dig more into the origin of this shape during runtime.

Reply:
0 in some dimensions is perfectly fine and should happen naturally during the course of many tf programs. For example, if to do partitioned embedding lookup, you take a tensor of indices and partition it so all indices in a range go to a certain partition. Since the content of the indices is dynamic, it's perfectly possible that a partition ends up with no indices, so shape [0]. Then if you use these indices to gather rows from a variable of shape [N, M], you get something with shape [0, M]. You can then apply a linear layer on these, and get something of the shape [0, K], etc.

If we cannot reason about these empty tensors with different shapes we're strongly limited on the type of shape dynamism we can support.

Reply:
Thanks @alextp  for the nice explanation!  @penpornk It seems the tensor shape with 0 in its dimension is a valid tensor, and I accidentally noticed some comments in c_api.cc where it created EmptyTensor, and did not care about shape.dims -- as long as the number of elements is 0. 

https://github.com/tensorflow/tensorflow/blob/0f08941cfbdf24474f7660cde5e880633a7e78be/tensorflow/c/c_api.cc#L390-L407

So during runtime, the above 0 sized tensor indeed can happen.

Reply:
Got it. So it's just a reserved shape without any elements. Thank you everyone for the explanation!

Reply:
Thank you @penpornk for reviewing and raising this question! 
BTW, I committed the strides change via github GUI. Please review :)